### ISSUE TITLE
The Affine model with MI is not satisfying for CT/MR co-registration

### ISSUE THREAD

**Issue by boveiri** [voxelmorph/voxelmorph#164]
Thank you in advance for your time and energy spent to maintain the issues updated!

I used the Affine model of voxelmorph with 8-bin Mutual-Information loss function for 2D CT/MR co-registration. The base-net architecture is as [32, 32, 32, 32, 16, 6] where 6 refers to the fully-connected last layer, and the other early layers are exactly like with the Affine model of voxelmorph. Here is some paired CT/MR slices:

http://s6.picofile.com/file/8392190192/myplot1.png

The model trained with adam (lr=1e-4) for 1000 epochs on 731 2D CT/MR pairs (80 ones for validation). The history is as follows:

http://s6.picofile.com/file/8392190242/MI8_ep1000.png

But, finally, the outcome was not satisfying both on the validation and test data, like the followings:

A validation data: http://s6.picofile.com/file/8392190226/1_val.png
A test data: http://s7.picofile.com/file/8392190234/2_test.png

I do not know why?! but the model with 'mse' loss function on 500 epochs was more satisfiable, though we are speaking about multimodal registration. I should note that both the train (val), and test data contain both T1 and T2 MRIs to make the problem more challenging, yet with image normalization we have done as a pre-processing, it should not be the source of problem, I am not sure. 

As a random guess, I am suspicious about MI loss with only 8 bins, and what do you think about 16 bins (unfortunately, the system cannot progress with more bins). 

Your help is highly appreciated. Thanks and regards.

**Comment by adalca** [comment]
@boveiri  I would definitely try more than 8 bins. Since it's only affine, 

I would suggest you make the images smaller (say downsize by half) and run with 16 or 32 bins (which might fit since the images will be much smaller). 

I also think that for the images you show, 3D would save you because the 3D shape of the brain and skull would disambiguate the rotation. 

I realize 3D might not fit in your system because of the 8-bin problem you mention above, but try to play with smaller data to start. Maybe if your volumes are 256^3, resize them to 64^3 and see if that works?

**Comment by boveiri** [comment]
Thank you for the information. The train on MSE was just completed for 1000 epochs, but I am not satisfied of the results, too. Here is the model fit history:

 http://s6.picofile.com/file/8392229726/MSE_ep1000.png

And, some following samples:

A validation sample: http://s7.picofile.com/file/8392229750/1_MSI_val.png
A test sample: http://s6.picofile.com/file/8392229826/1_MSI_test.png

I evaluated both of the models (MI-8bins and MSE), on the set of 278 paired slices as the test data, and the mean results are as follows:

MI-8bins Model: 
    Final loss on val. data: -0.0151 --- on the test data: -0.0125
    MSE metric on val: 0.764 --- on the test data: 0.702 (val. mse is worse than the test one!)

MSE Model:
    Final loss on val. data: 0.572 --- on the test data: 0.623
    MI-8bins on val data: -0.0164 --- on the test data: -0.0131

Based on your recommendation, I am increasing the no. of bins to 16 on down-sampled images and will report the results soon. For 3D registration, our data is unfortunately not completely compatible since it was gathered from a real clinical setup, not for research setup. Accordingly, the volumes are mostly (512, 512, 8) up to (512, 512, 12) where as can be seen the no. of slices in each vol. is a few (mostly because the expert annotators put some of the slices (first and last ones) away while annotating, and marked them as uninterested).

I have also the following questions and it is appreciated if you could help:

1- I normalize each image based on its own mean/std, where finally for all the data (test+train) we have:  
CT.min: -1.7617179 --- CT.max: 7.6282015
MR.min: -0.9354078 --- MR.max: 10.691616
Should I normalize the images based on the population mean/std? I mean separatedly for CTs and MRs as they are in different modalities.

2- Currently, we have not use any regularization loss for the Affine transformation output, where the grand-truth provided for the model is a np.zeros([6,]). I guess if we provided a MSE loss for it, it means that the produced transformation should be near (0, 0, 0, 0, 0, 0, 0) else will be punished and this avoids the network to select big values for the outputted transformation. Are you agree with this idea?

3- Why does the history diagrams so noisy even at the last epochs? I mean the network should be able to learn many in the early epochs and just fine-tune itself in the last ones, so that we should be witnessed of a smooth loss diagram for both the training and validation data especially in the last epochs!         

Thanks and regards,

**Comment by adalca** [comment]
HI @boveiri 

I don't expect MSE to do good things here, so I'd rather not comment on those results -- I think you should focus on the NMI. I also would encourage you *not* to look at your test set -- since you will be overfitting to it by looking at it. You need to keep a completely held out set until the end which you only test once!

Regarding the 3D: you could try to upsample to 4mm^3 (from probably 1x1x6 or so that your images are at right now) and get them all to be 64x64x64, and then apply the affine registration. I think at the end this is what you'll have to do, since 2D won't be practically useful (since the head itself will vary in the third dimension as well).

Regarding your questions:
1. I don't think it matters for NMI, as long as you provide the right bins to cover the range of the input images. I would actually normalize based on percentiles in each image (maybe 1 and 99 percentile)

2. I don't think providing a loss is necessary, but if you provide an MSE loss against a 0-vector, it's the equivalent of a gaussian prior on the parameters. This is reasonable to me if you think most of your deformations are small. It will be annoying to find the hyperparameter to balance this with the image loss.

3. This is fairly standard for SGD, especially with small batches -- since the batches are small, each image will create gradients that will affect the behaviour of the network significantly for the other images. If you want to see the smoothness, I recommend you use keras' [https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LearningRateScheduler](learning rate scheduler), maybe with a patience of 20 and going down by a factor of 0.5.

**Comment by boveiri** [comment]
Thanks for you informative comments. OK, hope I can finish this 2D version very fast to shift toward the 3D; however, my colleagues emphasis that interpolating more than 1 slice between two successive slices results in very ambitious and noisy images! so I think you mean replicating the slices several times to reach 64 slices? since resizing the volume from (512, 512, 12) to such ones like (64, 64, 64) is meaningless. If I am wrong please let me know.

Thanks for your answers on the questions, too. I just completed making the NCC model with 1000 epochs, and the fit history is as follows:

http://s6.picofile.com/file/8392287950/NCC_ep1000.png

It is wired as the NCC was going to increase instead of decrease! Also, the output of basenet that outputs the final affine transformation was gradually increased so that the basenet error reached to about 120,000 near the last epochs. On prediction, the output of model too high to be able to applied with the spatial-transformer layer, and actually there is no moved_image after the prediction!

http://s6.picofile.com/file/8392289576/1_val_NCC.png

**Comment by adalca** [comment]
@boveiri 

Interpolating by a large factor will certainly not give good interpolation results easily (although there is a *ton* of literature on how to do this well), but that's not the point -- you can simply interpolate linearly (even if the image doesn't look good) and then register. Remember, the goal is the registration (the affine parameters) not the warped image. 

Something is definitely wrong with the NCC loss, how are you computing that graph? I'm not sure what the 'basenet' is? Again, though, NCC is also not appropriate to MRI-CT registration, because it requires a monotonic relationship of tissue intensities.

**Comment by boveiri** [comment]
Hi Adrian,

I just finished training with MI-16bins. To save the time, I downscaled the images to (256, 256), and droop one of the convolutional layers out so that the network's structure was  [32, 32, 32, 16, 6] (with about 48,000 weights). I started with 500 epochs, but the final fit history showed the need for more epochs:

http://s7.picofile.com/file/8392535884/MI_16_Ep500.png

So, I continued with an extra set of 500 epochs, followed by a set of 1000 ones, totally constituting 2000 epochs (thanks to you for preparing get_config for the implementations so that e.g. the spatialtransformer layer and losses functions can be imported as custom objects).

http://s6.picofile.com/file/8392536142/1_val_MI_16_500_1000.png
http://s6.picofile.com/file/8392536168/MI_16_Ep1000_2000.png

Nevertheless, the final model was not satisfying e.g. on validation samples:

http://s7.picofile.com/file/8392536284/1_val_MI_16_1000_2000.png
http://s6.picofile.com/file/8392536318/2_val_MI_16_1000_2000.png

Do you think I need going for MI with more bins like 32? or the problem is elsewhere. As a guess, I think the 751 pairs are not sufficient for a good generalization. And, it may be fixed with a good data-augmentation and drop-out techniques.

I also took a look on some samples from the train data, and surprisingly there were a number of samples with poor align from the network, tough the metrics e.g. Dice or MSE shows some improvements (I did not checked the MI, but will do so after this post), the alignments were unacceptable, and can be far better. In some cases, the metrics shows worse alignment than the initial state! Such misalignment in the train data should have a sensible reason e.g. network simplicity, etc. 

Some samples:
http://s6.picofile.com/file/8392540200/1_Error_Train_MI_16_2000.png
http://s6.picofile.com/file/8392540268/2_Error_Train_MI_16_2000.png

*** The basenet is the sequential CNN model that outputs the transformation parameters (before the SpatialTransformer layer, which finally pack as a multi input/output keras API model). You named this sequential model as basenet in your implementations.

**Comment by boveiri** [comment]
Sorry, I know you are too busy to answer such trivial inquiries.

I am conducting a comprehensive set of experiments to investigate the impact of different influential factors e.g. no. of epochs, different losses, batch size, net. architecture and so on. 

1- I kindly want to know that why you used stride in conv. layer instead of using regular max-pooling layer? Do you have any background knowledge about it may be better in this specific case (registration, I mean)?

2- What is the impact of increasing the depth of either conv. layers or fully-connected ones? Does it have a tangible impact on the results?

**Comment by adalca** [comment]
Hi @boveiri sorry for the delay.

1. No real reason, we had to make a decision and in the latest literature for this sort of task there didn't seem to be much of a difference, but we never tested one or the other.

2. by depth do you mean the number of features? if so, increasing the number of features definitely helps accuracy in our experiments.

**Comment by boveiri** [comment]
Hi Adrian, happy to hearing from you. Thanks for your comments.

Actually, I found the source of error with try and error. The main cause of such a huge error rate was the batch-size! I set it to 1, and things went well. I think since the images are paired and each pair is independent of the others, they should not be mixed by batch-size more than 1. If there is other reasons so happy to know.

For my questions, I conducted some experiments and found that

1- As you told there was not any tangible increase in the performance when using stride=1 and max-pooling layers. Nevertheless, we added a dropout layer (0.3) after the flatten layer and it had a good contribution on the results. Also, we applied keras.regularizers.l2(0.0001) with about 0.1/0.2 contribution on the final loss, but it had about no impact!
        
2- we tried this architecture [64, 32, 32, 16, 6] on downsampled data to (256, 256), and it worked good. However, I think we need more experiments to find the best model. 

*** The experiments were conducted using MI with 64 bins. 

Meanwhile, I have some questions:

1- As the reception field is a decisive factor. Have you had any analysis on the receptive field of voxelmorph in any of your paper? Besides, manually calculation is there any formula to calculate the receptive field of a net. based on the no. of layers and size of the filters?

2- We want to know the MI of a pair of CT/MR images. Could you please tell us how it can be done using the MI function you written as a loss function. I tried the following, but did not work.

We changed the function parameters, and added two input parameters:
def globalMutualInformation1(y_true, y_pred, bin_centers, sigma_ratio=0.5, max_clip=1, crop_background=False): 

Also, the return was changed to: return loss (y_true, y_pred)

Suppose that the images are in two Tensors: timgCT and timgMR, also the bin centers are in an ndarray named bin_centers. We used the following command, but got the following error. We appreciate if you could help us in this regard. 

>>> sess.run(vxm.losses.globalMutualInformation1(timgCT, timgMR, bin_centers))

FailedPreconditionError (see above for traceback): Attempting to use uninitialized value Variable_5
	 [[node Variable_5/read (defined at C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:402)  = Identity[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Variable_5)]]
	 [[{{node Neg_2/_1}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_66_Neg_2", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

**Comment by happybcloud** [comment]
my input size is [batchsize,256,256,256,4], the src size is [batchsize,256,256,256,2],the tgt size is [batchsize,256,256,256,2],can I still get a flow with size [256,256,256,3] that can warp src to y using nrn_layers.SpatialTransformer

**Comment by adalca** [comment]
@happybcloud thanks for the feeback

@boveiri  sorry for missing this earlier.

1 - we haven't really analyzed receptive field, to be honest. In our experience, simply going 4-5 layers should do the trick in most situations. The full formula would depend on the number of convolution layers and the number of pooling levels -- for each pooling layer you double your receptive field, and each conv adds a couple of voxels per level depending on the kernel size. In simple terms it's roughly 2^5 or so voxels?

2. Sorry, I don't understand why you changed the inputs. The mutual information losses do exactly what you want. For example setting `localMutualInformation(np.linspace(0,1,32), vol_size).local_mi` as your loss should do the trick, assuming you normalized both of your images to [0, 1].

**Comment by happybcloud** [comment]
thanks for the feeback,.Maybe I got it wrong. I need to learn more

**Comment by happybcloud** [comment]
> @happybcloud thanks for the feeback
> 
> @boveiri sorry for missing this earlier.
> 
> 1 - we haven't really analyzed receptive field, to be honest. In our experience, simply going 4-5 layers should do the trick in most situations. The full formula would depend on the number of convolution layers and the number of pooling levels -- for each pooling layer you double your receptive field, and each conv adds a couple of voxels per level depending on the kernel size. In simple terms it's roughly 2^5 or so voxels?
> 
> 1. Sorry, I don't understand why you changed the inputs. The mutual information losses do exactly what you want. For example setting `localMutualInformation(np.linspace(0,1,32), vol_size).local_mi` as your loss should do the trick, assuming you normalized both of your images to [0, 1].
thanks for the feeback,.Maybe I got it wrong. I need to learn more

**Comment by happybcloud** [comment]
thanks for the feeback,.Maybe I got it wrong. I need to learn more
